libcst/codegen/gen_visitor_functions.py

```diff
@@ -0,0 +1,163 @@
+for node, base in nodebases.items():
+    if node.__name__.startswith("Base"):
+        continue
+    for x in (node, base):
+        imports[x.__module__].add(x.__name__)
+
+generated_code: List[str] = []
+generated_code.append(
+    "# This source code is licensed under the MIT license found in the"
+)
+generated_code.append("# LICENSE file in the root directory of this source tree.")
+generated_code.append("")
+generated_code.append("# pyre-strict")
+generated_code.append("")
+generated_code.append("# This file was generated by libcst.codegen.gen_matcher_classes")
+generated_code.append("from typing import Optional, Union, TYPE_CHECKING")
+generated_code.append("")
+generated_code.append("from libcst._removal_sentinel import RemovalSentinel")
+generated_code.append("from libcst._typed_visitor_base import mark_no_op")
+
+# Import the types we use. These have to be type guarded since it would
+# cause an import cycle otherwise.
+generated_code.append("")
+generated_code.append("")
+generated_code.append(f"if TYPE_CHECKING:")
+for module, objects in imports.items():
+    generated_code.append(f"    from {module} import (  # noqa: F401")
+    generated_code.append(f"        {', '.join(sorted(list(objects)))}")
+    generated_code.append("    )")
+
+# Generate the base visit_ methods
+generated_code.append("")
+generated_code.append("")
+generated_code.append("class CSTTypedBaseFunctions:")
+for node in sorted(nodebases.keys(), key=lambda node: node.__name__):
+    name = node.__name__
+    if name.startswith("Base"):
+        continue
+    generated_code.append("")
+    generated_code.append("    @mark_no_op")
+    generated_code.append(
+        f'    def visit_{name}(self, node: "{name}") -> Optional[bool]:'
+    )
+    generated_code.append("        pass")
+
+# Generate the visitor leave_ methods
+generated_code.append("")
+generated_code.append("")
+generated_code.append("class CSTTypedVisitorFunctions(CSTTypedBaseFunctions):")
+for node in sorted(nodebases.keys(), key=lambda node: node.__name__):
+    name = node.__name__
+    if name.startswith("Base"):
+        continue
+    generated_code.append("")
+    generated_code.append("    @mark_no_op")
+    generated_code.append(
+        f'    def leave_{name}(self, original_node: "{name}") -> None:'
+    )
+    generated_code.append("        pass")
+
+# Generate the transformer leave_ methods
+generated_code.append("")
+generated_code.append("")
+generated_code.append("class CSTTypedTransformerFunctions(CSTTypedBaseFunctions):")
+generated_code.append("    pass")
+for node in sorted(nodebases.keys(), key=lambda node: node.__name__):
+    name = node.__name__
+    if name.startswith("Base"):
+        continue
+    generated_code.append("")
+    generated_code.append("    @mark_no_op")
+    generated_code.append(
+        f'    def leave_{name}(self, original_node: "{name}", updated_node: "{name}") -> Union["{nodebases[node].__name__}", RemovalSentinel]:'
```

Comment:
While this is mostly correct, I noticed that there are a few cases where we accept a union of multiple potential values. For example, I might want to replace a compound statement with a simple statement, so the return type should actually be `Union["SimpleStatementLine", "BaseCompoundStatement"]`.

Or maybe I want to replace every comma with a `MaybeSentinel`.

I'm not sure that we should type the return value beyond `Union[CSTNode, MaybeSentinel, RemovalSentinel]`.

Reply:
That's not actually correct though. If you replace a simple statement with a compound statement, you have no guarantee that the node isn't a child of a SimpleStatementLine.

Reply:
Going all the way to allowing `CSTNode` would mean losing type safety in cases where you should be returning an expression but the user tries to return a statement. Looking briefly at the generated code and how inheritance is done with nodes it looks like its mostly an issue with statements as there's a lot of disjoint base classes for various kinds of statements.

Reply:
Yes, but I think it's always valid to replace a `BaseCompoundStatement` with a `SimpleStatementLine`. I'd rather have the wider type because it's more functional even if it's not always as safe.

Reply:
I did some more thinking on this: Essentially, typing the response as CSTNode throws away the benefit of bothering to generate these functions in the first place. We're trying to generate functions that give a user the most constrained version available so that its less likely that invalid code is generated. Stating that all leave_* nodes can return this is not just wrong (not many node parents support a MaybeSentinel, and nobody needs to return a MaybeSentinel instead of a conrete type), but also useless from a typing perspective. If that's the case, we literally do not need to generate code since the whole purpose is for typing.

Reply:
Even if the return type is `CSTNode`, it's still useful in ensuring that the arguments match the function name. Lint AST visitors never had return types, but we still codegen'ed them for this reason.

We could still check the return types using `validate_types_shallow` on the resulting parent node. In fact, we should probably default to running `validate_types_shallow` in `__post_init__` for people using LibCST without mypy or pyre.